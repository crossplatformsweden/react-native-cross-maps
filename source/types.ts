// map/types/index

/**
 * Map action types
 * @constant
 * @public
 * @typedef MapConstants
 * @type {Object}
 */
export enum MapConstants {
  MAP_LOCATION_BUSY = 'Letar upp dig',
  MAP_LOCATION_SUCCESS = 'MAP = GET LOCATION SUCCESS',
  MAP_LOCATION_FAILED = 'MAP = GET LOCATION FAILED',
}

/**
 * Describes coordiantes for a location on the map
 * @typedef ILocation
 * @type {Object}
 * @public
 * @export
 * @property {number?} latitude
 * @property {number?} longitude
 * @public
 */
export interface ILatLng {
  latitude: number;
  longitude: number;
}

/**
 * Describes an area on the map
 *
 * @export
 * @interface ILocation
 * @extends {ILatLng}
 * @property {number?} latitudeDelta
 * @property {number?} longitudeDelta
 */
export interface IRegion {
  latitude: number;
  longitude: number;
  latitudeDelta: number;
  longitudeDelta: number;
}

export interface MapViewDirectionsProps {
  origin?: Array<any>;
  waypoints?: Array<any>;
  destination?: Array<any>;
  apikey: string;
  onStart?(): void;
  onReady?(): void;
  onError?(): void;
  mode?: 'DRIVING' | 'BICYCLING' | 'TRANSIT' | 'WALKING';
  language?: string;
  resetOnChange?: boolean;
  optimizeWaypoints?: boolean;
  directionsServiceBaseUrl?: string;
  region?: string;
}

export interface ILocation extends ILatLng {
  altitude: Number;
  timestamp: Number; // Milliseconds since Unix epoch
  accuracy: Number;
  altitudeAccuracy: Number;
  speed: Number;
}

/**
 * Describes padding on the map
 *
 * @export
 * @interface IEdgePadding
 */
export interface IEdgePadding {
  top: Number;
  right: Number;
  bottom: Number;
  left: Number;
}

/**
 * Describes a pin on the map
 *
 * @export
 * @interface IMapPin
 */

export interface IMapPin {
  isUser: boolean;
  location: string;
  title?: string;
  description?: string;
}

/**
 * Reduced properties passed down from LoginReducer
 * @typedef IMapState
 * @export
 * @public
 * @property {ILocation} location
 * @property {boolean} isUserSearch
 */
export interface IMapState {
  location: IRegion;
  /**
   * determines if the location changed from user search. Default is false
   *
   * @type {boolean}
   * @memberof IMapState
   */
  isUserSearch: boolean;
  __typename: MapConstants;
}

/**
 * Creates a new {IMapState}
 * @param location user location
 * @param isUserSearch determines if the location changed from user search. Default is false
 * @param type action type
 */
export const MapState = (
  location: IRegion = {
    latitude: 0,
    longitude: 0,
    latitudeDelta: 0,
    longitudeDelta: 0,
  },
  isUserSearch = false,
  __typename = MapConstants.MAP_LOCATION_BUSY
) => {
  const r: IMapState = {
    location,
    isUserSearch,
    __typename,
  };
  return r;
};
