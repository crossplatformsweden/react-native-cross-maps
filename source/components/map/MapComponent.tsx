import * as React from 'react';
import _ from 'lodash';
import { View } from 'react-native';
// @ts-ignore - missing in definition
import MapView, { PROVIDER_GOOGLE } from 'react-native-maps';
import * as types from '../../types';
import LocationService from '../../services/location-service/LocationService';
import * as MapService from '../../services/map-service/MapService';
import { ILatLng } from '../../types';
// @ts-ignore
import MapViewDirections from 'react-native-maps-directions';
import UserLocationButton from '../../components/user-location-button/UserLocationButton';
import { styles, Colors } from 'react-native-cross-components';
import MapViewProps from 'react-native-maps';
import { MapViewDirectionsProps, IRegion } from '../../types';

interface IState {
  mapMargin: number;
  userLocation?: ILatLng;
}

interface IProps {
  mapViewProps?: MapViewProps;
  mapNavigationProps?: MapViewDirectionsProps;
  initialRegion?: IRegion;
  destination?: ILatLng | null;
  apikey?: string;
  userLocationButton?: boolean;
  goToLocation?(newLocation: types.ILatLng): void;
  onRegionChange?: (newRegion: IRegion) => void;
  onUserLocationChanged?: (newLocation: ILatLng) => void;
}

let map: any;

export class MapComponent extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);
    this.state = {
      mapMargin: 1,
    };

    // this.setMap = this.setMap.bind(this);
    this.getPredefinedPlaces = this.getPredefinedPlaces.bind(this);
    this.goToUserLocation = this.goToUserLocation.bind(this);
    this.onMapReady = this.onMapReady.bind(this);
  }

  public async goToLocation(coordinate: types.ILatLng) {
    if (!map) {
      return;
    }

    MapService.goToUserLocation(map, coordinate);
  }

  public async goToUserLocation() {
    console.log('goToUserLocation has map? ' + !_.isNil(map));
    if (!map) {
      return;
    }

    console.log(
      'goToUserLocation has LocationService? ' + !_.isNil(LocationService)
    );
    const coordinate = await LocationService.getCurrentLocation();
    MapService.goToUserLocation(map, coordinate);
    this.setState({ userLocation: coordinate });
    if (this.props.onUserLocationChanged) {
      console.log('goToUserLocation calls onUserLocationChanged');
      this.props.onUserLocationChanged(coordinate);
    }
  }

  /**
   * Get list of predefined places available in autocomplete
   */
  getPredefinedPlaces(): Array<Object> {
    // const { userLocation } = this.state;
    // if (!_.isNil(userLocation)) {
    //   const userPlace = {
    //     description: 'Min nuvarande plats',
    //     geometry: {
    //       location: {
    //         lat: userLocation.latitude,
    //         lng: userLocation.longitude,
    //       },
    //     },
    //   };
    //   return [userPlace];
    // }

    return [];
  }

  setMap(ref: any) {
    map = ref;
  }

  onMapReady() {
    this.setState({ mapMargin: 0 });
  }

  render() {
    return (
      <View style={styles.container}>
        <MapView
          maxZoomLevel={18}
          zoomEnabled={true}
          loadingEnabled={false}
          showsBuildings={false}
          showsUserLocation
          showsIndoors={false}
          showsMyLocationButton={false}
          toolbarEnabled={false}
          onMapReady={this.onMapReady}
          onRegionChangeComplete={(region: types.IRegion) => {
            if (!_.isNil(this.props.onRegionChange)) {
              this.props.onRegionChange(region);
            }
          }}
          loadingIndicatorColor={Colors.CrossLightBlue}
          showsCompass={false}
          ref={this.setMap}
          style={[styles.container, { marginBottom: this.state.mapMargin }]}
          provider={PROVIDER_GOOGLE}
          {...this.props.mapViewProps}
        >
          {this.state.userLocation &&
          this.props.destination &&
          this.props.destination.latitude ? (
            <MapViewDirections
              origin={this.state.userLocation}
              destination={{
                latitude: Number(this.props.destination.latitude),
                longitude: Number(this.props.destination.longitude),
              }}
              //TODO PARAMS
              apikey={this.props.apikey}
              strokeColor="blue"
              strokeWidth={3}
              {...this.props.mapNavigationProps}
            />
          ) : null}
          {this.props.children}
        </MapView>
        {this.props.userLocationButton ? (
          <UserLocationButton onPress={this.goToUserLocation} />
        ) : null}
      </View>
    );
  }
}

export default MapComponent;
