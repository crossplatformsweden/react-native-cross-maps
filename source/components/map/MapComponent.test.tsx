import React from 'react';
import MapComponent from './MapComponent';
import TestRenderer from 'react-test-renderer';
import { ILatLng } from '../../types';

jest.unmock('react-native');

const coords = {
  latitude: 1,
  longitude: 0,
};

// Default export requires this type of mocking
jest.mock('react-native-maps', () => ({
  __esModule: true,
  MapView: 'View',
  Marker: 'View',
  Callout: 'View',
  Ploygon: 'View',
  Polyline: 'View',
  Circle: 'View',
  Overlay: 'View',
  PROVIDER_GOOGLE: '',
  MarkerAnimated: 'View',
  default: 'View',
}));

jest.mock('../../services/location-service/LocationService', () => ({
  getCurrentLocation: () =>
    new Promise<ILatLng>((resolve) => {
      resolve({
        latitude: 3,
        longitude: 2,
      });
    }),
}));

jest.mock('../../services/map-service/MapService', () => ({
  goToUserLocation: () => console.log('IM A MOCKED GOTOUSERLOCATON FUNC'),
}));

function setup() {
  return TestRenderer.create(<MapComponent />);
}

describe('components', () => {
  describe('<MapComponent />', () => {
    it('renders correctly with defaults', () => {
      const wrapper = setup();
      expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('Should have state.mapMargin = 1', () => {
      const wrapper = setup();
      const state = wrapper.root.instance.state;
      expect(state.mapMargin).toBe(1);
    });

    it('api key', () => {
      const wrapper = TestRenderer.create(<MapComponent apikey="1234" />);
      expect(wrapper.root.instance.props.apikey).toBe('1234');
    });

    it('destination prop can be inserted by user', () => {
      const wrapper = TestRenderer.create(
        <MapComponent destination={coords} />
      );
      expect(wrapper.root.instance.props.destination).toBe(coords);
    });

    // it('onRegionChange is called', () => {
    //   const wrapper = TestRenderer.create(<MapComponent />);
    //   expect(wrapper.root.props.instance.onRegionChange).toBeCalled();
    // });

    it('Go to user location should set state', async () => {
      let res;
      const func = (data: any) => {
        console.log('HERE', data);
        res = data;
      };
      const wrapper = TestRenderer.create(
        <MapComponent onUserLocationChanged={func} />
      );
      const inst: any = wrapper.getInstance();
      if (!inst) {
        throw new Error('No instance');
      }
      inst.setMap(1);
      await inst.goToUserLocation();
      console.log('FINNIDHED EXECUTING');
      // expect(func).toBeCalled();

      expect(res).toBeDefined();
      // const state = wrapper.root.instance.state;
      // console.log('userLocation', state.userLocation);
      // expect(state.userLocation).toBeDefined();
      //expect(func).toBeCalledTimes(1);

      expect(res).toEqual({
        latitude: 3,
        longitude: 2,
      });
    }, 8000);

    it('MapView has "maxZoomLevel" = "18"', () => {
      const wrapper = setup();
      const child = wrapper.root.findByProps({ maxZoomLevel: 18 });
      expect(child).toBeDefined();
    });
  });
});
