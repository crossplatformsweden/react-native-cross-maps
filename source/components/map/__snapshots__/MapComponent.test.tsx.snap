// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components <MapComponent /> renders correctly with defaults 1`] = `
<View
  style={Object {}}
>
  <View
    loadingEnabled={false}
    loadingIndicatorColor="#057dc2"
    maxZoomLevel={18}
    onMapReady={[Function]}
    onRegionChangeComplete={[Function]}
    provider=""
    showsBuildings={false}
    showsCompass={false}
    showsIndoors={false}
    showsMyLocationButton={false}
    showsUserLocation={true}
    style={
      Array [
        Object {},
        Object {
          "marginBottom": 1,
        },
      ]
    }
    toolbarEnabled={false}
    zoomEnabled={true}
  />
</View>
`;
